x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/indices/recovery/SearchableSnapshotRecoveryState.java

```diff
@@ -0,0 +1,115 @@
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.common.Nullable;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public final class SearchableSnapshotRecoveryState extends RecoveryState {
+    private boolean preWarmFinished;
+
+    public SearchableSnapshotRecoveryState(ShardRouting shardRouting, DiscoveryNode targetNode, @Nullable DiscoveryNode sourceNode) {
+        super(shardRouting, targetNode, sourceNode, new Index());
+    }
+
+    @Override
+    public synchronized RecoveryState setStage(Stage stage) {
+        // The transition to the final state was done by #prewarmCompleted, just ignore the transition
+        if (getStage() == Stage.DONE) {
+            return this;
+        }
+
+        // Pre-warm is still running, hold the state transition
+        // until the pre-warm process finishes
+        if (preWarmFinished == false && stage == Stage.DONE) {
+            return this;
+        }
+
+        return super.setStage(stage);
+    }
+
+    public synchronized void preWarmFinished() {
+        // For small shards it's possible that the
+        // cache is pre-warmed before the stage has transitioned
+        // to FINALIZE, so the transition to the final state is delayed until
+        // the recovery process catches up.
+        if (getStage() == Stage.FINALIZE) {
+            super.setStage(Stage.DONE);
+        }
+
+        SearchableSnapshotRecoveryState.Index index = (Index) getIndex();
+        index.stopTimer();
+        preWarmFinished = true;
+    }
+
+    public synchronized void ignoreFile(String name) {
+        SearchableSnapshotRecoveryState.Index index = (Index) getIndex();
+        index.addFileToIgnore(name);
+    }
+
+    private static final class Index extends RecoveryState.Index {
+        // We ignore the files that won't be part of the pre-warming
+        // phase since the information for those files won't be
+        // updated and marking them as reused might be confusing,
+        // as they are fetched on-demand from the underlying repository.
+        private final Set<String> filesToIgnore = new HashSet<>();
+
+        private Index() {
+            super(new SearchableSnapshotRecoveryFilesDetails());
+            // We start loading data just at the beginning
+            super.start();
+        }
+
+        private synchronized void addFileToIgnore(String name) {
+            filesToIgnore.add(name);
+        }
+
+        @Override
+        public synchronized void addFileDetail(String name, long length, boolean reused) {
+            if (filesToIgnore.contains(name)) {
+                return;
+            }
+
+            super.addFileDetail(name, length, reused);
+        }
+
+        // We have to bypass all the calls to the timer
+        @Override
+        public synchronized void start() {}
+
+        @Override
+        public synchronized void stop() {}
+
+        @Override
+        public synchronized void reset() {}
```

Comment:
do we not need to implement this method, which  gets called when a peer recovery retries?

Reply:
The behavior don't change during peer recovery, the cache starts pre-warming as soon as the directory is created if I'm not mistaken.